src/main/java/org/myrobotlab/service/ServoMixer.java

```diff
@@ -9,75 +9,188 @@
 import org.slf4j.Logger;
 
 /**
  * ServoMixer - a service which can control multiple servos. The position of the
  * servos can be saved into multiple poses, and the poses in turn can be saved
- * in a sequence file to play back later.
+ * in a gesture file to play back later.
+ * 
+ * @author GroG, kwatters, others...
+ * 
  */
-public class ServoMixer extends Service {
+public class ServoMixer extends Service implements ServiceLifeCycleListener {
 
-  public final static Logger log = LoggerFactory.getLogger(InMoov2.class);
+  public class Player implements Runnable {
+    String gestureName = null;
+    int poseIndex = 0;
+    boolean running = false;
+    Gesture runningGesture = null;
+    Thread thread = null;
 
-  private static final long serialVersionUID = 1L;
+    private void play() {
+      if (runningGesture.parts != null) {
+        invoke("publishGestureStarted", gestureName);
+        for (int i = 0; i < runningGesture.parts.size(); ++i) {
+          if (!running) {
+            break;
+          }
+          GesturePart sp = runningGesture.parts.get(i);
+          invoke("publishPlayingGesturePart", sp);
+          invoke("publishPlayingGesturePartIndex", i);
+          switch(sp.type) {
+            case "Pose":{
+              Pose pose = getPose(sp.name);
+              if (pose == null) {
+                warn("Pose %s not found", sp.name);
+                continue;
+              }          
+              // move to positions
+              moveToPose(sp.name, pose, sp.blocking);
+              // TODO if stopped, get and stop all servos
+
+            }
+            break;
+            case "Delay":{
+                sleep((Integer)sp.value);
+            }
+            break;
+            case "Speech":{
+              invoke("publishText", (String)sp.value);
+            }
+          break;            
+            default:{
+              error("do not know how to handle gesture part of type %s", sp.type);
+            }
+          }
+        } // poses
+        
+        invoke("publishGestureStopped", gestureName);
 
-  protected String servoMixerDirectory = getDataDir() + fs + "poses";
+      }
+      running = false;
+    }
 
-  /**
-   * set autoDisable on "all" servos .. true - will make all servos autoDisable
-   * false - will make all servos autoDisable false null - will make no changes
-   */
-  protected Boolean autoDisable = null;;
+    @Override
+    public void run() {
+      try {
+        running = true;
+        if (runningGesture.repeat) {
+          while (running) {
+            play();
+          }
+        } else {
+          play();
+        }
+      } catch (Exception e) {
+        error(e);
+      }
+      running = false;
+    }
+
+    public void start(String name, Gesture seq) {
+      gestureName = name;
+      runningGesture = seq;
+      poseIndex++;
+      thread = new Thread(this, String.format("%s-player-%d", getName(), poseIndex));

```

Comment:
Should probably use an executor pool of some kind, creating a thread for every gesture is expensive. Also, I take it this is not just a subclass of a thread so that it can run multiple gestures simultaneously? If so, there needs to be thread locals or concurrent maps for `running` and other shared state

Reply:
yup .. updated

Reply:
no .. not multiple gestures simultaneously .. that's why lots of the old python gestures scripts have finishGesture() calls at the end .... in that they want to lock & unlock a gesture ...
It just needs a single thread ... I find defining it as Runnable allows more flexibility, including moving it to an executor pool .. which i've done (should at least cut down on the startup time...)